### PR TITLE
MNIST GAN Model and Trainer interfaces

### PR DESCRIPTION
**Pull request by q349wang**
This PR consists of several topics:

- Generic DCGAN model for 32x32 grayscale images
- Data Generator interfaces that create `tf.data.Dataset` objects
- Trainer interface for training of GAN model
- MNIST data generator class for use in training
- Fashion MNIST data generator class for use in training
- MNIST trainer which can use either of the above mentioned generators for training a GAN

Some results:
Fashion MNIST 100 epochs
FAKEgenerated_plot_e10003d

MNIST 50 epochs
FAKEgenerated_plot_e5003d

### CODE REVIEW

File: model/gan/gan.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        # Convolution2D Transpose 2
+        model.add(layers.Conv2DTranspose(64 * self._NUM_CHANNELS, (4, 4),
+                                         strides=(2, 2), padding='same', use_bias=False))
+        assert model.output_shape == (None, 16, 16, 64 * self._NUM_CHANNELS)
+        model.add(layers.BatchNormalization(momentum=self._MOMENTUM))
+        model.add(layers.LeakyReLU(alpha=self._ALPHA))
+
+        # Convolution2D Transpose 3
+        # model.add(layers.Conv2DTranspose(32 * self._NUM_CHANNELS, (4, 4),
```
Comment (t6george) [inline model/gan/gan.py:67]: Do you want to leave this in for experimentation later?

File: model/datagen/fashionmnist_datagen.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    def load_data(self, batch:int) -> (tf.data.Dataset, int):
+        """
+        Get dataset as array of batches
+
+        @params batch: Number of data elements in batch
+        @returns Dataset representing data as well as the number of elements in an epoch
+        """
+        
+        return self.dataset.cache().shuffle(self.info.splits['train'].num_examples).repeat().batch(batch, drop_remainder=True).prefetch(tf.data.experimental.AUTOTUNE), self.info.splits['train'].num_examples
```
Comment (t6george) [inline model/datagen/fashionmnist_datagen.py:31]: This line is hella long, cs343 submit command would not accept. This is really minor though, if you want we can fix it down the road.

File: model/datagen/mnist_datagen.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def load_data(self, batch:int) -> (tf.data.Dataset, int):
+        """
+        Get dataset as array of batches
+
+        @params batch: Number of data elements in batch
+        @returns Dataset representing data as well as the number of elements in an epoch
+        """
+        
+        return self.mnist_dataset.cache().shuffle(self.info.splits['train'].num_examples).repeat().batch(batch, drop_remainder=True).prefetch(tf.data.experimental.AUTOTUNE), self.info.splits['train'].num_examples
```
Comment (t6george) [inline model/datagen/mnist_datagen.py:31]: Long line here too - can fix now or later

File: model/trainers/mnist_trainer.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        # Default training protocol
+        half_batch = int(batch / 2)
+        for i in range(epoch):
+            for j in range(num_batches):
+                x_real, y_real = self.generate_real_samples(half_batch)
+
+                x_fake, y_fake = self.generate_fake_samples(half_batch)
+
+                x, y = np.vstack((x_real, x_fake)), np.vstack((y_real, y_fake))
```
Comment (t6george) [inline model/trainers/mnist_trainer.py:122]: Should these be shuffled?

**Review by t6george — APPROVED** [review]
Sound and Complete

File: model/gan/gan.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                                padding='same'))
+        model.add(layers.LeakyReLU(alpha=self._ALPHA))
+        model.add(layers.Dropout(0.4))
+
+        # Convolution layer 3
+        # model.add(layers.Conv2D(64, (3, 3), strides=(2, 2),
+        #                         padding='same'))
+        # model.add(layers.LeakyReLU(alpha=self._ALPHA))
+        # model.add(layers.Dropout(0.4))
+
```
[the remark this answers is not in the document]
Comment (q349wang) [reply model/gan/gan.py:102]: nah we will be going back to that when we have more compute power/time

[the remark this answers is not in the document]
Comment (q349wang) [reply model/gan/gan.py:102]: Will add a comment

File: model/gan/gan.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        assert model.output_shape == (None, 16, 16, 64 * self._NUM_CHANNELS)
+        model.add(layers.BatchNormalization(momentum=self._MOMENTUM))
+        model.add(layers.LeakyReLU(alpha=self._ALPHA))
+
+        # Convolution2D Transpose 3
+        # model.add(layers.Conv2DTranspose(32 * self._NUM_CHANNELS, (4, 4),
+        #                                  strides=(2, 2), padding='same', use_bias=False))
+        # # assert model.output_shape == (None, 64, 64, 32 * self._NUM_CHANNELS)
+        # model.add(layers.BatchNormalization(momentum=self._MOMENTUM))
+        # model.add(layers.LeakyReLU(alpha=self._ALPHA))
```
[the remark this answers is not in the document]
Comment (q349wang) [reply model/gan/gan.py:71]: see [https://github.com/q349wang/moss/pull/6#discussion_r513169011](url)

File: model/trainers/mnist_trainer.py (hunk 4 above)
Comment (q349wang) [reply model/trainers/mnist_trainer.py:122]: I dont think its necessary to shuffle them. either way the discriminator is being trained on them

Comment (q349wang) [reply model/trainers/mnist_trainer.py:122]: @Taha-Masood to confirm

Comment (t6george) [reply model/trainers/mnist_trainer.py:122]: Ok yeah just making sure that the model doesn't detect a pattern in the input

**Review by erikterwiel — APPROVED** [review]
LGTM 👍

File: model/gan/gan.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        """
+        """
+        try:
+
+            g_name = F"./model_saves/generator_model_{epoch + 1}.h5"
+            d_name = F"./model_saves/discriminator_model_{epoch + 1}.h5"
+            gan_name = F"./model_saves/gan_model_{epoch + 1}.h5"
+
+            self._discriminator.trainable = False
+            self._gan.save(gan_name)
```
Comment (Taha-Masood) [inline model/gan/gan.py:130]: Pretty sure this will default to save using the TF SavedModel format not HDF5, however, your file name seems to suggest otherwise (https://www.tensorflow.org/api_docs/python/tf/keras/Model#save).

File: model/trainers/mnist_trainer.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+from model.trainers.trainer import trainer
+from model.gan.gan import gan
+
+from model.datagen.datagen import datagen
+from model.datagen.mnist_datagen import mnist_datagen
+from model.datagen.fashionmnist_datagen import fashionmnist_datagen
+
+class mnist_trainer(trainer):
+
+    BATCH_SIZE = 64
```
Comment (Taha-Masood) [inline model/trainers/mnist_trainer.py:15]: During the next mp we can try and clean up the hyper-parameters (ideally moving them out of the class and having them as configuration variables). That way we could grid-search over the hyper-parameters easily.

File: model/trainers/trainer.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+                d_loss, _ = gan_model.get_discriminator().train_on_batch(x, y)
+
+                x_gen = self.generate_noise(gan_model._noise_dim, batch)
+                y_gen = np.ones((batch, 1))
+
+                g_loss = gan_model.get_gan().train_on_batch(x_gen, y_gen)
+
+                # summarize loss on this batch
+                print('>%d, %d/%d, d=%.3f, g=%.3f' % (i+1, j+1, num_batches, d_loss, g_loss))
```
Comment (Taha-Masood) [inline model/trainers/trainer.py:146]: We are kind of inconsistent on using F strings and old Python style strings. Perhaps we can mark this as a TODO for mp3.

File: model/gan/gan.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
+import tensorflow as tf
+import numpy as np
+import logging
+
+from tensorflow.keras import layers
+
+
+class gan():
+    _ALPHA = 0.25
```
Comment (Taha-Masood) [inline model/gan/gan.py:9]: Same as the other hyper-parameter comment.

**Review by Taha-Masood — APPROVED** [review]
LGTM (comments are just notes for future work).
